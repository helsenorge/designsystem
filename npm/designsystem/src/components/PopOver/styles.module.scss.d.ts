export type Styles = {
  anchorlink: string;
  anchorlink__icon: string;
  'anchorlink-wrapper': string;
  popover: string;
  popover__arrow: string;
  'popover__arrow--over': string;
  'popover__arrow--under': string;
  'popover__arrow--visible': string;
  'popover--visible': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
