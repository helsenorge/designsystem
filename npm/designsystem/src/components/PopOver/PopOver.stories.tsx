import React, { useRef } from 'react';

import { action } from '@storybook/addon-actions';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import PopOver, { PopOverVariant } from './PopOver';
import { useToggle } from '../../hooks/useToggle';
import loremText, { longLoremText } from '../../utils/loremtext';
import Button from '../Button';
import GridExample from '../GridExample';
import Icon from '../Icons';
import HelpSign from '../Icons/HelpSign';
import Table, { TableHead, TableRow, TableHeadCell, TableBody, TableCell, ResponsiveTableVariant } from '../Table';

export default {
  title: '@helsenorge∕designsystem-react/Components/PopOver',
  component: PopOver,
  parameters: {
    docs: {
      description: {
        component: 'PopOver er en liten popup som lar innbygger lese et tekstinnhold som utdyper det som ble trykket på.',
      },
    },
  },
  argTypes: {
    children: {
      control: 'text',
      defaultValue:
        'Dette er en PopOver. Aliquip aute consectetur eiusmod nisi ullamco aliquip adipisicing cupidatat reprehenderit nulla in Lorem sint.',
    },
    showBubble: {
      control: 'boolean',
      defaultValue: true,
    },
    variant: {
      control: 'select',
      options: PopOverVariant,
      defaultValue: PopOverVariant.positionautomatic,
    },
  },
} as ComponentMeta<typeof PopOver>;

export const Default: ComponentStory<typeof PopOver> = (args: any) => {
  const controllerRef = useRef<SVGSVGElement>(null);

  return (
    <GridExample>
      <span>{loremText + loremText + loremText + loremText}</span>
      <div style={{ position: 'relative', display: 'inline' }}>
        <Icon ref={controllerRef} svgIcon={HelpSign} />
        <PopOver {...args} onLinkClick={action('Mer hjelp clicked')} onClose={action('Bubble closed')} controllerRef={controllerRef}>
          <div style={{ padding: '0.5rem 1rem' }}>{args.children}</div>
        </PopOver>
      </div>
      <span>{loremText + loremText + loremText + loremText}</span>
    </GridExample>
  );
};

export const Toggle: ComponentStory<typeof PopOver> = (args: any) => {
  const controllerRef = useRef<HTMLButtonElement>(null);
  const { value, toggleValue } = useToggle(false);

  return (
    <GridExample>
      <span>{loremText + loremText + loremText + loremText}</span>
      <div style={{ position: 'relative', display: 'inline' }}>
        <Button ref={controllerRef} onClick={toggleValue}>
          {'Åpne'}
        </Button>
        {value && (
          <PopOver {...args} onLinkClick={action('Mer hjelp clicked')} onClose={action('Bubble closed')} controllerRef={controllerRef}>
            <div style={{ padding: '0.5rem 1rem' }}>{args.children}</div>
          </PopOver>
        )}
      </div>
      <span>{loremText + loremText + loremText + loremText}</span>
    </GridExample>
  );
};

export const HorizontalScroll: ComponentStory<typeof Table> = (args: any) => {
  const controllerRef = useRef<SVGSVGElement>(null);

  return (
    <GridExample>
      <p>{longLoremText}</p>
      <Table breakpointConfig={{ breakpoint: 'xl', variant: ResponsiveTableVariant.horizontalscroll }}>
        <TableHead category={args.headerCategory}>
          <TableRow key="head">
            <TableHeadCell>Fastlege</TableHeadCell>
            <TableHeadCell>Fastlegekontor</TableHeadCell>
            <TableHeadCell>Ledige plasser</TableHeadCell>
            <TableHeadCell>Antall på venteliste</TableHeadCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell dataLabel="navn">Line Danser</TableCell>
            <TableCell dataLabel="kontor">Røtvedt</TableCell>
            <TableCell dataLabel="ledige">1</TableCell>
            <TableCell dataLabel="antall">
              <Icon ref={controllerRef} svgIcon={HelpSign} />
              <PopOver {...args} onClose={action('Bubble closed')} controllerRef={controllerRef}>
                <div style={{ padding: '0.5rem 1rem' }}>{args.children}</div>
              </PopOver>
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell dataLabel="navn">Line Danser</TableCell>
            <TableCell dataLabel="kontor">Røtvedt</TableCell>
            <TableCell dataLabel="ledige">1</TableCell>
            <TableCell dataLabel="antall">200</TableCell>
          </TableRow>
          <TableRow>
            <TableCell dataLabel="navn">Line Danser</TableCell>
            <TableCell dataLabel="kontor">Røtvedt</TableCell>
            <TableCell dataLabel="ledige">1</TableCell>
            <TableCell dataLabel="antall">200</TableCell>
          </TableRow>
          <TableRow>
            <TableCell dataLabel="navn">Line Danser</TableCell>
            <TableCell dataLabel="kontor">Røtvedt</TableCell>
            <TableCell dataLabel="ledige">1</TableCell>
            <TableCell dataLabel="antall">200</TableCell>
          </TableRow>
        </TableBody>
      </Table>
      <p>{longLoremText}</p>
    </GridExample>
  );
};
