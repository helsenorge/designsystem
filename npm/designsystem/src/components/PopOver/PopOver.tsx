import React, { useEffect, useRef, useState } from 'react';

import classNames from 'classnames';

import { getArrowStyle, getBubbleStyle, getVerticalPosition } from './utils';
import { AnalyticsId, ZIndex } from '../../constants';
import { useInterval } from '../../hooks/useInterval';
import { useIsVisible } from '../../hooks/useIsVisible';
import { useLayoutEvent } from '../../hooks/useLayoutEvent';
import { useSize } from '../../hooks/useSize';
import { mergeRefs } from '../../utils/refs';

import styles from './styles.module.scss';

export enum PopOverVariant {
  positionautomatic = 'positionautomatic',
  positionbelow = 'positionbelow',
  positionabove = 'positionabove',
}

export type PopOverRole = 'tooltip';

export interface PopOverProps {
  /** Id of the PopOver */
  id?: string;
  /** Content shown inside PopOver. Note that if role="tooltip", you must not include interactive/focusable elements. */
  children: React.ReactNode;
  /** Ref for the element the PopOver is placed upon */
  controllerRef: React.RefObject<HTMLElement | SVGSVGElement>;
  /** Ref for the element the PopOver is placed upon */
  popOverRef?: React.RefObject<HTMLDivElement>;
  /** Show the popover. Only applies when role=tooltip. Default: false. */
  show?: boolean;
  /** Adds custom classes to the element. */
  className?: string;
  /** Adds custom classes to the arrow element. */
  arrowClassName?: string;
  /** Determines the placement of the popover. Default: automatic positioning. */
  variant?: keyof typeof PopOverVariant;
  /** Sets role of the PopOver element */
  role?: PopOverRole;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Overrides the default z-index of PopOver */
  zIndex?: number;
}

const PopOver = React.forwardRef<HTMLDivElement | SVGSVGElement, PopOverProps>((props, ref) => {
  const {
    id,
    children,
    controllerRef,
    popOverRef,
    show = false,
    className = '',
    variant = PopOverVariant.positionautomatic,
    role,
    testId,
    arrowClassName,
    zIndex = ZIndex.PopOver,
  } = props;

  const bubbleRef = popOverRef || useRef<HTMLDivElement>(null);
  const arrowRef = useRef<HTMLDivElement>(null);
  const bubbleSize = useSize(bubbleRef);
  const [controllerSize, setControllerSize] = useState<DOMRect>();
  const controllerisVisible = useIsVisible(controllerRef, 0);

  const updateControllerSize = (): void => {
    setControllerSize(controllerRef.current?.getBoundingClientRect());
  };

  useInterval(updateControllerSize, 500);
  useLayoutEvent(updateControllerSize, ['scroll', 'resize'], 10);

  useEffect(() => {
    updateControllerSize();
  }, []);

  const isTooltip = role === 'tooltip';

  const popOverClasses = classNames(styles.popover, { [styles['popover--visible']]: isTooltip ? show : controllerisVisible }, className);
  const verticalPosition = controllerSize && bubbleSize && getVerticalPosition(controllerSize, bubbleSize, variant);
  const arrowClasses = classNames(styles.popover__arrow, arrowClassName, {
    [styles['popover__arrow--over']]: verticalPosition === PopOverVariant.positionbelow,
    [styles['popover__arrow--under']]: verticalPosition === PopOverVariant.positionabove,
    [styles['popover__arrow--visible']]: isTooltip ? show : controllerisVisible,
  });

  const bubbleStyle = controllerSize && bubbleSize && getBubbleStyle(controllerSize, bubbleSize, variant);
  const arrowStyle = bubbleStyle && controllerSize && verticalPosition && getArrowStyle(bubbleStyle, controllerSize, verticalPosition);

  return (
    <>
      <div
        id={id}
        ref={mergeRefs([ref, bubbleRef])}
        className={popOverClasses}
        style={{ ...bubbleStyle, zIndex }}
        data-testid={testId}
        data-analyticsid={AnalyticsId.PopOver}
        role={role}
      >
        {children}
      </div>
      <div ref={arrowRef} className={arrowClasses} style={{ ...arrowStyle, zIndex }} />
    </>
  );
});

PopOver.displayName = 'PopOver';

export default PopOver;
