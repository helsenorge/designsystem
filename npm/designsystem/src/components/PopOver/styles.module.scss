@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';
@import '../AnchorLink/styles.module';

.popover {
  $popover: &;

  display: flex;
  position: fixed;
  align-items: flex-start !important;
  max-width: 23.3125rem;
  text-align: start;
  font-family: inherit;
  font-size: $font-size-sm;
  line-height: $lineheight-size-sm;
  background-color: $white;
  z-index: 3;
  visibility: hidden;

  &--visible {
    visibility: visible;
  }

  border: getSpacer(4xs) solid $plum600;
  border-radius: 0.5625rem;
  box-shadow: 0 0.125rem 1.125rem 0 rgb(0 0 0 / 15%);

  @media (min-width: map.get($grid-breakpoints, lg)) {
    font-size: $font-size-md;
    line-height: $lineheight-size-md;
  }

  &__arrow {
    --drop-shadow-color: #{$plum600};

    width: 0;
    height: 0;
    position: fixed;
    border-style: solid;
    border-color: transparent;
    z-index: 3;
    visibility: hidden;

    &--over {
      border-width: 0.625rem;
      border-bottom-color: $white;
      filter: drop-shadow(var(--drop-shadow-color) 0 0.125rem 0.375rem 0.125rem);
      filter: drop-shadow(0 -0.185rem 0 var(--drop-shadow-color));
    }

    &--under {
      border-width: 0.625rem;
      border-top-color: $white;
      filter: drop-shadow(var(--drop-shadow-color) 0 0.125rem 0.375rem 0.125rem);
      filter: drop-shadow(0 0.2rem 0 var(--drop-shadow-color));
    }

    &--visible {
      visibility: visible;
    }
  }
}
