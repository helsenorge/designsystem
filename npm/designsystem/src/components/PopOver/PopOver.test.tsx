import React, { useRef } from 'react';

import { screen, render, waitFor } from '@testing-library/react';

import PopOver, { PopOverProps, PopOverVariant } from './PopOver';

const { dummyDomRect } = vi.hoisted(() => {
  return {
    dummyDomRect: {
      height: 100,
      width: 100,
      top: 0,
      left: 0,
      right: 100,
      bottom: 100,
    },
  };
});

vi.mock('../../hooks/useSize', () => ({
  useSize: vi.fn().mockReturnValue(dummyDomRect),
}));
vi.mock('../../hooks/useIsVisible', () => ({
  useIsVisible: vi.fn().mockReturnValue(true),
}));

const PopOverWithController: React.FC<Omit<PopOverProps, 'controllerRef'>> = props => {
  const controllerRef = useRef<HTMLSpanElement>(null);

  return (
    <>
      <span ref={controllerRef}>{'Test'}</span>
      <PopOver {...props} controllerRef={controllerRef} />
    </>
  );
};

describe('Gitt at PopOver skal vises', (): void => {
  beforeAll(() => {
    window.HTMLDivElement.prototype.getBoundingClientRect = vi.fn().mockReturnValue(dummyDomRect);
  });

  describe('Når den skal vises vanlig', (): void => {
    it('Så vises PopOver som vanlig', (): void => {
      const { container } = render(<PopOverWithController testId="test01">{'Test tekst'}</PopOverWithController>);

      const bubble = screen.getByTestId('test01');
      const child = screen.getByText('Test tekst');

      expect(bubble).toBeInTheDocument();
      expect(child).toBeInTheDocument();

      expect(bubble).toHaveClass('popover');

      expect(container).toMatchSnapshot();
    });
  });

  describe('Når variant er positionbelow', (): void => {
    it('Så vises PopOver riktig', async (): Promise<void> => {
      render(
        <PopOverWithController variant={PopOverVariant.positionbelow} testId="test01">
          {'Test tekst'}
        </PopOverWithController>
      );

      const bubble = screen.getByTestId('test01');
      // eslint-disable-next-line testing-library/no-node-access
      const arrow = bubble.nextSibling;

      expect(bubble).toHaveClass('popover');
      await waitFor(() => expect(arrow).toHaveClass('popover__arrow popover__arrow--over'));
    });
  });

  describe('Når variant er positionabove', (): void => {
    it('Så vises PopOver riktig', async (): Promise<void> => {
      render(
        <PopOverWithController variant={PopOverVariant.positionabove} testId="test01">
          {'Test tekst'}
        </PopOverWithController>
      );

      const bubble = screen.getByTestId('test01');
      // eslint-disable-next-line testing-library/no-node-access
      const arrow = bubble.nextSibling;

      expect(bubble).toHaveClass('popover');
      await waitFor(() => expect(arrow).toHaveClass('popover__arrow popover__arrow--under'));
    });
  });
});
