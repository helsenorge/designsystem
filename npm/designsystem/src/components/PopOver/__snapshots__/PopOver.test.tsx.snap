// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at PopOver skal vises > Når den skal vises vanlig > Så vises PopOver som vanlig 1`] = `
<div>
  <span>
    Test
  </span>
  <div
    class="popover popover--visible"
    data-analyticsid="pop-over"
    data-testid="test01"
    style="z-index: 11000; right: 12px; top: 16px; width: -24px;"
  >
    Test tekst
  </div>
  <div
    class="popover__arrow popover__arrow--over popover__arrow--visible"
    style="z-index: 11000; right: 24px; top: 0px;"
  />
</div>
`;
